cockatrice/src/gamesmodel.cpp

```diff
@@ -257,30 +119,28 @@ QVariant GamesModel::headerData(int section, Qt::Orientation /*orientation*/, in
     }
 }
 
-const ServerInfo_Game &GamesModel::getGame(int row)
+const GameListItem &GamesModel::getGame(int row)
 {
     Q_ASSERT(row < gameList.size());
-    return gameList[row];
+    return gameMap.find(gameList[row]).value();
 }
 
-void GamesModel::updateGameList(const ServerInfo_Game &game)
+void GamesModel::rebuildList()
 {
-    for (int i = 0; i < gameList.size(); i++) {
-        if (gameList[i].game_id() == game.game_id()) {
-            if (game.closed()) {
-                beginRemoveRows(QModelIndex(), i, i);
-                gameList.removeAt(i);
-                endRemoveRows();
-            } else {
-                gameList[i].MergeFrom(game);
-                emit dataChanged(index(i, 0), index(i, NUM_COLS - 1));
-            }
-            return;
-        }
+    beginResetModel();
+    gameList.clear();
+    for (auto &game : gameMap.values()) {
+        addGame(game);
     }
-    beginInsertRows(QModelIndex(), gameList.size(), gameList.size());
-    gameList.append(game);
-    endInsertRows();
+    endResetModel();
+}
+
+void GamesModel::addGame(GameListItem &game)
+{
+    gameList.append(game.getId());
+    int row = gameList.size() - 1;
+    game.updateFunc([this, row]() { emit dataChanged(index(row, 0), index(row, NUM_COLS - 1)); });
```

Comment:
Please add a comment here, this is funky

Reply:
I just went with this in the end because I wanted it to compile before asking you for your opinion on the changes to where the list of games is stored

Reply:
So this adds a list (well qmap) for gamesbyid; I am working on making games, replays and decks openable from outside the client. So links in the form `cockatrice://example.eu?roomID=123&gameID=321&password=123remind%20me%20to%20escape%20special%20chars` as this will make my tournament bot have far nicer interfacing with users.